cupyx/scipy/interpolate/_polyint.py

```diff
@@ -0,0 +1,323 @@
+import cupy
+from cupyx.scipy._lib._util import _asarray_validated
+
+
+class _Interpolator1D:
+    """Common features in univariate interpolation.
+
+    Deal with input data type and interpolation axis rolling. The
+    actual interpolator can assume the y-data is of shape (n, r) where
+    `n` is the number of x-points, and `r` the number of variables,
+    and use self.dtype as the y-data type.
+
+    Attributes
+    ----------
+    _y_axis : Axis along which the interpolation goes in the
+        original array
+    _y_extra_shape : Additional shape of the input arrays, excluding
+        the interpolation axis
+    dtype : Dtype of the y-data arrays. It can be set via _set_dtype,
+        which forces it to be float or complex
+
+    Methods
+    -------
+    __call__
+    _prepare_x
+    _finish_y
+    _reshape_y
+    _reshape_yi
+    _set_yi
+    _set_dtype
+    _evaluate
+
+    """
+
+    def __init__(self, xi=None, yi=None, axis=None):
+        self._y_axis = axis
+        self._y_extra_shape = None
+        self.dtype = None
+        if yi is not None:
+            self._set_yi(yi, xi=xi, axis=axis)
+
+    def __call__(self, x):
+        """Evaluate the interpolant
+
+        Parametres
+        ----------
+        x : array-like
+            The points to evaluate the interpolant
+
+        Returns
+        -------
+        y : array-like
+            Interpolated values. Shape is determined by replacing
+            the interpolation axis in the original array with the shape of x
+
+        Notes
+        -----
+        Input values `x` must be convertible to `float` values like `int`
+        or `float`.
+
+        """
+        x, x_shape = self._prepare_x(x)
+        y = self._evaluate(x)
+        return self._finish_y(y, x_shape)
+
+    def _evaluate(self, x):
+        """
+        Actually evaluate the value of the interpolator
+        """
+        raise NotImplementedError()
+
+    def _prepare_x(self, x):
+        """
+        Reshape input array to 1-D
+        """
+        x = _asarray_validated(x, check_finite=False, as_inexact=True)
+        x_shape = x.shape
+        return x.ravel(), x_shape
+
+    def _finish_y(self, y, x_shape):
+        """
+        Reshape interpolated y back to an N-D array similar to initial y
+        """
+        y = y.reshape(x_shape + self._y_extra_shape)
+        if self._y_axis != 0 and x_shape != ():
+            nx = len(x_shape)
+            ny = len(self._y_extra_shape)
+            s = (list(range(nx, nx + self._y_axis))
+                 + list(range(nx)) + list(range(nx + self._y_axis, nx + ny)))
+            y = y.transpose(s)
+        return y
+
+    def _reshape_yi(self, yi, check=False):
+        """
+        Reshape the updated yi to a 1-D array
+        """
+        yi = cupy.moveaxis(cupy.asarray(yi), self._y_axis, 0)
```

Comment:
Let's remove `cupy.array` to avoid implicit numpy to cupy conversion.
```suggestion
        yi = cupy.moveaxis(yi, self._y_axis, 0)
```

Reply:
Got it. Thanks!